cpp/src/arrow/type_traits.h

```diff
@@ -41,102 +41,78 @@ struct TypeTraits<NullType> {
-struct TypeTraits<Int16Type> {
-  using ArrayType = Int16Array;
-  using BuilderType = Int16Builder;
-  using TensorType = Int16Tensor;
-
-  static constexpr int64_t bytes_required(int64_t elements) {
-    return elements * sizeof(int16_t);
-  }
-  constexpr static bool is_parameter_free = true;
-  static inline std::shared_ptr<DataType> type_singleton() { return int16(); }
-};
-
-template <>
-struct TypeTraits<UInt32Type> {
-  using ArrayType = UInt32Array;
-  using BuilderType = UInt32Builder;
-  using TensorType = UInt32Tensor;
-
-  static constexpr int64_t bytes_required(int64_t elements) {
-    return elements * sizeof(uint32_t);
-  }
-  constexpr static bool is_parameter_free = true;
-  static inline std::shared_ptr<DataType> type_singleton() { return uint32(); }
-};
-
-template <>
-struct TypeTraits<Int32Type> {
-  using ArrayType = Int32Array;
-  using BuilderType = Int32Builder;
-  using TensorType = Int32Tensor;
+struct TypeTraits<BooleanType> {
+  using ArrayType = BooleanArray;
+  using BuilderType = BooleanBuilder;
+  using CType = bool;
 
   static constexpr int64_t bytes_required(int64_t elements) {
-    return elements * sizeof(int32_t);
+    return BitUtil::BytesForBits(elements);
   }
   constexpr static bool is_parameter_free = true;
-  static inline std::shared_ptr<DataType> type_singleton() { return int32(); }
+  static inline std::shared_ptr<DataType> type_singleton() { return boolean(); }
 };
 
 template <>
-struct TypeTraits<UInt64Type> {
-  using ArrayType = UInt64Array;
-  using BuilderType = UInt64Builder;
-  using TensorType = UInt64Tensor;
+struct TypeTraits<bool> {
+  using ArrayType = BooleanArray;
+  using BuilderType = BooleanBuilder;
+  using ArrowType = BooleanType;
 
   static constexpr int64_t bytes_required(int64_t elements) {
-    return elements * sizeof(uint64_t);
+    return BitUtil::BytesForBits(elements);
   }
   constexpr static bool is_parameter_free = true;
-  static inline std::shared_ptr<DataType> type_singleton() { return uint64(); }
+  static inline std::shared_ptr<DataType> type_singleton() { return boolean(); }
 };
 
-template <>
-struct TypeTraits<Int64Type> {
-  using ArrayType = Int64Array;
-  using BuilderType = Int64Builder;
-  using TensorType = Int64Tensor;
+#define PRIMITIVE_TYPE_TRAITS_DEF_(CType_, ArrowType_, ArrowArrayType, ArrowBuilderType, \
+                                   ArrowTensorType, SingletonFn)                       \
+  template <>                                                                            \
+  struct TypeTraits<ArrowType_> {                                                        \
+    using ArrayType = ArrowArrayType;                                                    \
+    using BuilderType = ArrowBuilderType;                                                \
+    using TensorType = ArrowTensorType;                                                  \
+    using CType = CType_;                                                                \
```

Comment:
We use `c_type` in `CTypeImpl::c_type`. Either way we're inconsistent, though ;-)

Reply:
Which inconsistency should we prefer? I prefer being close to the other using declaration in the same struct.

Reply:
I have no strong opinion on this.